### PR TITLE
Introduce callconv annotation

### LINKED ISSUE #661: use a builtin enum for calling conventions instead of keywords
right now we have these keywords for function calling conventions:
 * `extern` - must have C ABI. Also used to indicate C ABI on structs, enums, and unions.
 * `stdcallcc`
 * `coldcc` - cold calling convention is a thing, but this might be better off as an optimizer hint that the function is cold rather than specifying a calling convention. Currently the line is blurred.
 * `nakedcc` - used for entry points like `_start` and when inline assembly defines the whole function.

If no calling convention is specified, then the calling convention is selected automatically (and currently defaults to the "fast calling convention" that LLVM has available.) Related issue: #105 when mixing automatic calling convention with specified calling conventions.
Any function can be `export`ed, and if the calling convention of an exported function is not specified, then `extern` is chosen.

This proposal is to make a builtin enum:

```zig
pub const CC = enum {
    Auto,
    C,
    StdCall,
    Cold,
    Naked,
};
```

And instead of keywords, the `fn` keyword has optional calling convention specifier that matches `align`, like this:

```zig
export fn _start() callconv(builtin.CC.Naked) {
    // ...
}
```

If a calling convention is not specified, `builtin.CC.Auto` is the default. This is similar to the endianness in pointers proposal (see #649). When displaying a fn type, the calling convention only has to be specified if it is not `builtin.CC.Auto`.

Then it becomes easier to add reflection with `@typeInfo(@typeOf(_start))`. `builtin.TypeInfo.CallingConvention` should be pulled out into `builtin.CallingConvention`

Now that enum literals have arrived, this proposal is nicer:

```zig
export fn _start() callconv(.Naked) {
    // ...
}
```

### PR DESCRIPTION
**Pull request by LemonBoy**
Done according to the accepted proposal, `zig fmt` has been patched up to use the new notation when possible and the whole stdlib has been already updated.
A few new calling conventions have been introduced, you can now write your interrupt handlers (for x86 & AVR) with pure Zig code.

Missing:
- [x] Integrate with `translate-c-2`

Not critical but possibly wanted:
- [ ] Validate `export`/`extern`/`callconv` combinations
- [ ] Checks to prevent the use of unsupported CCs for a given target arch

A few open questions:
- Some CCs are architecture specific, do we want to add a prefix to those (eg. `Fastcall` -> `X86_Fastcall`) ?
- Right now we "group" all the CCs under the same name (eg. AVR interrupt & X86 interrupt are both under the `Interrupt` entry) and are chosen according to the target architecture. Is this ok or do we want to split them into independent entities?

I'm submitting this as-is to see if the CI is green, ~in case this is merged **please** rebase it~. 

Closes #661

### CODE REVIEW

File: lib/std/mutex.zig [hunk 1]
```diff
@@ -48,8 +48,8 @@ pub const Mutex = if (builtin.single_threaded)
             return self.tryAcquire() orelse @panic("deadlock detected");
         }
     }
-else if (builtin.os == .windows) 
-    // https://locklessinc.com/articles/keyed_events/
+else if (builtin.os == .windows)
+// https://locklessinc.com/articles/keyed_events/
```
Comment (daurnimator) [inline lib/std/mutex.zig]: This doesn't look right to me. Created https://github.com/ziglang/zig/issues/3978

File: src/analyze.cpp [hunk 2]
```diff
@@ -1442,8 +1449,8 @@ ZigType *analyze_type_expr(CodeGen *g, Scope *scope, AstNode *node) {
 ZigType *get_generic_fn_type(CodeGen *g, FnTypeId *fn_type_id) {
     ZigType *fn_type = new_type_table_entry(ZigTypeIdFn);
     buf_resize(&fn_type->name, 0);
-    const char *cc_str = calling_convention_fn_type_str(fn_type->data.fn.fn_type_id.cc);
-    buf_appendf(&fn_type->name, "%s", cc_str);
+    if (fn_type->data.fn.fn_type_id.cc == CallingConventionC)
+        buf_append_str(&fn_type->name, "extern ");
```
Comment (andrewrk) [inline src/analyze.cpp]: I was thinking we could drop `extern` for the C calling convention, once we have `callconv`. Any reason to keep using `extern` for C calling convention functions?

File: src/all_types.hpp [hunk 3]
```diff
@@ -639,10 +648,14 @@ struct AstNodeFnProto {
...
     AstNode *section_expr;
+    // populated if the "callconv(S)" is present
+    AstNode *callconv_expr;
     Buf doc_comments;
 
     FnInline fn_inline;
-    CallingConvention cc;
+    bool is_nakedcc;
+    bool is_stdcallcc;
+    bool is_async;
```
Comment (andrewrk) [inline src/all_types.hpp]: Feel free to delete nakedcc and stdcallcc keywords from stage1. (If you don't do it, I'll do it.) I'm happy to tell people to use the auto-fixing you added to `zig fmt` in order to upgrade code to the latest version.

I'll open a separate proposal to remove support for `async fn` in favor of `callconv(.Async)` for uniformity.

File: lib/std/zig/render.zig [hunk 4]
```diff
@@ -1295,8 +1295,16 @@ fn renderExpression(
                 try renderExpression(allocator, stream, tree, indent, start_col, lib_name, Space.Space);
             }
 
+            // Some extra machinery is needed to rewrite the old-style cc
+            // notation to the new callconv one
```
Comment (andrewrk) [inline lib/std/zig/render.zig]: Thanks for adding this auto-upgrading functionality to zig fmt, it's quite nice. Would you mind adding a `testTransform` to the top of parser_test.zig along with a TODO, to remind us to remove this functionality after 1 release cycle?

File: src/analyze.cpp [hunk 5]
```diff
@@ -2162,7 +2180,7 @@ static Error resolve_struct_type(CodeGen *g, ZigType *struct_type) {
             ZigType *field_type = resolve_struct_field_type(g, field);
             if (field_type == nullptr) {
                 struct_type->data.structure.resolve_status = ResolveStatusInvalid;
-                return err;
+                return ErrorSemanticAnalyzeFail;
```
Comment (andrewrk) [inline src/analyze.cpp]: yikes! nice catch.

File: src/codegen.cpp [hunk 6]
```diff
@@ -285,14 +285,54 @@ static LLVMCallConv get_llvm_cc(CodeGen *g, CallingConvention cc) {
...
-            }
+            return LLVMCCallConv;
+        case CallingConventionFastcall:
+            if (g->zig_target->arch == ZigLLVM_x86)
+                return LLVMX86FastcallCallConv;
+            return LLVMCCallConv;
+        case CallingConventionVectorcall:
+            if (g->zig_target->arch == ZigLLVM_x86)
+                return LLVMX86VectorCallCallConv;
+        // XXX Enable this when the C API exports this enum member too
```
Comment (andrewrk) [inline src/codegen.cpp]: We should work around this with `zig_llvm.h`/`zig_llvm.cpp`. The official LLVM C API always lags behind the C++ API.

In fact: 

https://github.com/ziglang/zig/blob/86ba8c06bff84ac4865cef18080fa64fe946cb11/src/zig_llvm.h#L74-L75

https://github.com/ziglang/zig/blob/86ba8c06bff84ac4865cef18080fa64fe946cb11/src/zig_llvm.cpp#L275-L297

Looks like all we're missing is a copy-paste of LLVM's calling convention enum, with some `static_assert`s in zig_llvm.cpp to make sure the values don't change. See src/zig_clang.cpp for inspiration.

File: src/analyze.cpp (hunk 2 above)
Comment (LemonBoy) [reply src/analyze.cpp]: Decoupling `extern` from the CC is definitely a good idea, I tried not to change the status quo too much in this PR so I think I'll leave the breaking changes to you :)

File: src/all_types.hpp (hunk 3 above)
Comment (LemonBoy) [reply src/all_types.hpp]: Ditto, if you want to drop `nakedcc` and `stdcallcc` in this release cycle go for it! The less legacy cruft the better.

File: lib/std/zig/render.zig (hunk 4 above)
Comment (LemonBoy) [reply lib/std/zig/render.zig]: Will do!

File: src/codegen.cpp (hunk 6 above)
Comment (LemonBoy) [reply src/codegen.cpp]: I've [sent](https://reviews.llvm.org/D71942) a patch that hopefully will get in by the time LLVM 10 is out. But moving away from the C API is also a good idea.

**Comment by andrewrk** [comment]
> IMO it's fine to leave the enums un-prefixed and merged with some hints in the documentation about what to expect.

:+1:

**Comment by LemonBoy** [comment]
This is ready to go on my end, I'll send a follow-up PR to make `compiler-rt` use the AAPCS cc for the AEABI builtins once this is merged.

**Pull request merged**